File: Python/96.Unique Binary Search Trees.py
#写出有n个数字的二叉搜索树，一共有多少种可能的形状
#0个数的共1种，空树
#1个数的共1种，1个结点
#2个数的共2种，1做根和2做根各一种，1*1+1*1
#3个数的共5种，1做根的2种，2做根的1种，3做根的2种，1*2+1*1+2*1
#..........
#n个数的情况，每个数都可以做根，设第n个数共有X[n]种，那么X[n] = X[0]*X[n-1]+X[1]*X[n-2]+...+X[n-1]*X[0]
class Solution(object):
    def numTrees(self, n):
        if n==0:
            return 1
        X = [0 for i in range(n+1)]
        X[0] , X[1] = 1 , 1
        for i in range(2,n+1):
            for j in range(i):
                X[i] += X[j] * X[i-j-1]
        return X[n]


s = Solution()
print(s.numTrees(0))

File: C++/Tree/96.Unique Binary Search Trees.cpp
/*
给出正整数n，问有n个结点的二叉搜索树有多少种?
例如：给出3，返回5。
   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
*/
/*
思路：设G(n)为最终要求的结果，n个结点二叉树个数。
	  设F(i,n)为当i为二叉树根结点时的n个结点的二叉树个数。
	  由定义可知，G(n)=F(1,n)+F(2,n)+..+F(n,n)。也就是分别把每个数当做根结点的总和。
	  G(0) = 1， G(1) = 1.
	  此时，我们分析当i为根结点时，左侧有i-1个结点，右侧有n-i个结点。当根结点为i时的二叉树个数
	  就是左侧i-1个结点能构成的个数，乘上右侧的。也就是F(i,n) = G(i-1)*G(n-i)。
	  所以最终的式子为G(n) = G(0)*G(n-1)+G(1)*G(n-2)+..+G(n-1)G(0)。
	  也就是G(n)依赖于G(0)...G(n-1)的值。
 */
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int numTrees(int n) {
        vector<int> G(n+1);
        G[0] = G[1] = 1;
        for(int i=2; i <= n; ++i)
        {
        	for(int j=1; j <=i; ++j)
        		G[i] += G[j-1] * G[i-j];
        }
        return G[n];
    }
};
int main()
{
	Solution s;
	cout << s.numTrees() <<endl;
}

